src/python/pants/engine/internals/graph_test.py

```diff
@@ -83,13 +102,257 @@ def get_target(name: str) -> Target:
+
+        This is a tricky edge case to trigger. First, we must be using the original owning targets,
+        rather than generated subtargets, which means that there must be multiple owning targets.
+        Then, we must have two specs that resulted in the same original address.
+        """
+        self.create_files("demo", ["f1.txt", "f2.txt"])
+        self.add_to_build_file(
+            "demo",
+            dedent(
+                """\
+                target(name='one', sources=['*.txt'])
+                target(name='two', sources=['*.txt'])
+                """
+            ),
+        )
+        specs = [FilesystemLiteralSpec("demo/f1.txt"), FilesystemLiteralSpec("demo/f2.txt")]
+        result = self.request_single_product(
+            AddressesWithOrigins, Params(FilesystemSpecs(specs), create_options_bootstrapper())
+        )
+        expected_origin = FilesystemMergedSpec.create(specs)
+        assert result == AddressesWithOrigins(
+            [
+                AddressWithOrigin(Address("demo", "two"), expected_origin),
+                AddressWithOrigin(Address("demo", "one"), expected_origin),
+            ]
         )
+
+    def test_filesystem_specs_nonexistent_file(self) -> None:
+        specs = FilesystemSpecs([FilesystemLiteralSpec("demo/fake.txt")])
+        with pytest.raises(ExecutionError) as exc:
+            self.request_single_product(
+                AddressesWithOrigins, Params(specs, create_options_bootstrapper()),
+            )
+        assert 'Unmatched glob from file arguments: "demo/fake.txt"' in str(exc.value)
+        ignore_errors_result = self.request_single_product(
+            AddressesWithOrigins,
+            Params(specs, create_options_bootstrapper(args=["--owners-not-found-behavior=ignore"])),
+        )
+        assert not ignore_errors_result
+
+    def test_filesystem_specs_no_owner(self) -> None:
+        self.create_file("no_owners/f.txt")
+        # Error for literal specs.
+        with pytest.raises(ExecutionError) as exc:
+            self.request_single_product(
+                AddressesWithOrigins,
+                Params(
+                    FilesystemSpecs([FilesystemLiteralSpec("no_owners/f.txt")]),
+                    create_options_bootstrapper(),
+                ),
+            )
+        assert "No owning targets could be found for the file `no_owners/f.txt`" in str(exc.value)
+
+        # Do not error for glob specs.
+        glob_file_result = self.request_single_product(
+            AddressesWithOrigins,
+            Params(
+                FilesystemSpecs([FilesystemGlobSpec("no_owners/*.txt")]),
+                create_options_bootstrapper(),
+            ),
+        )
+        assert not glob_file_result
+
+    def test_resolve_addresses(self) -> None:
+        """This tests that we correctly merge addresses resolved from address specs with those
+        resolved from filesystem specs.
```

Comment:
It's worth considering whether we really want to do this. IIRC, during the design discussion there seemed to be consensus that if you ran `./pants test '**/*'` it was ok to run "test per file". Perhaps there have been other discoveries since then that suggest that that would not be a good idea, but it would seem to be much easier, and possibly more consistent?

Reply:
> It's worth considering whether we really want to do this. IIRC, during the design discussion there seemed to be consensus that if you ran ./pants test '**/*' it was ok to run "test per file"... ie, for subtarget addresses to be used. Perhaps there have been other discoveries since then that suggest that that would not be a good idea, but it would seem to be much easier, and possibly more consistent?

I don't follow. We would end up running "test per file", except if there were multiple owners for that file. This test is only what to do if you run `./pants test '**/*' ::`. In that case, the `::` would "win out" over the generated subtargets from the file args. This is simply a test for the behavior added in https://github.com/pantsbuild/pants/pull/10345.

> I know that we're trying to defer the "multiple owning targets" syntax, but if it avoids this type of special casing, it might be worth it.

I disagree on it being a priority for Pants 2.0. The special casing is complex, but it's well isolated to `internals/graph.py` and it's well tested. If it were leaking throughout the codebase, I would think differently.

In terms of special-casing being complex to users, I think we teach this that "When it's safe to do so, Pants will try to use a more precise target than you put in your BUILD file. Otherwise, it will use the coarser target you explicitly defined."